### PR TITLE
Upgrade to NXP release 5.4.24 2.1.0

### PR DESCRIPTION
**Pull request by zandrey**
This is a follow-up of the #456 and it brings the first round of recipes which are upgraded to NXP release `upgrade-to-rel_imx_5.4.24_2.1.0` tag.

This PR contains base recipes, which are mainly reside in `recipes-bsp` and `recipes-kernel`.

So far, following components are updated:
```
kernel-module-qca9377:     update recipe name to 3.1
kernel-module-imx-gpu-viv: upgrade to 6.4.0.p2.4
alsa-state:                add support for wm8960 codec
firmware-qca:              upgrade to 3.0.2
linux-imx-headers:         upgrade to rel_imx_5.4.24_2.1.0
linux-fslc-imx:            upgrade to rel_imx_5.4.24_2.1.0 from NXP
imx-seco:                  upgrade to version 3.6.3
imx-mkimage:               upgrade to version 1.0
imx-sc-firmware:           upgrade to 1.5.0 from NXP
imx-atf:                   upgrade to 2.2
u-boot-imx-tools:          upgrade to 2020.04
u-boot-imx-mfgtool:        upgrade to 2020.04
u-boot-imx:                upgrade to imx_v2020.04_5.4.24_2.1.0
```

This upgrade has been boot-tested on *imx8mmevk* and *imx8mnevk* HW, both booted into command prompt.

I've opened this PR as a draft PR for now, so it gets a chance to be looked at by community and I would also be adding few recipes here as well.

-- andrey

### CODE REVIEW

File: recipes-bsp/u-boot/u-boot-imx-common.inc [hunk 1]
```diff
@@ -3,9 +3,9 @@ DESCRIPTION = "i.MX U-Boot suppporting i.MX reference boards."
 LICENSE = "GPLv2+"
 LIC_FILES_CHKSUM = "file://Licenses/gpl-2.0.txt;md5=b234ee4d69f5fce4486a80fdaf4a4263"
 
-SRCBRANCH = "lf-5.4.y_v2019.04"
+SRCBRANCH = "imx_v2020.04_5.4.24_2.1.0"
 SRC_URI = "git://source.codeaurora.org/external/imx/uboot-imx.git;protocol=https;branch=${SRCBRANCH}"
-SRCREV = "228843cdf5435d4bd69f42a6015f78761ff4cc0d"
+SRCREV = "4979a99482f7e04a3c1f4fb55e3182395ee8f710"
```
Comment (otavio) [inline recipes-bsp/u-boot/u-boot-imx-common.inc:8]: The `u-boot-imx-tools` and `u-boot-imx-mfgtools` ought to be upgraded together. So please squash the following two commits into this one.

File: recipes-bsp/imx-sc-firmware/imx-sc-firmware_1.5.0.bb [hunk 2]
```diff
@@ -1,22 +1,26 @@
 # Copyright (C) 2016 Freescale Semiconductor
-# Copyright (C) 2017-2019 NXP
+# Copyright 2017-2019 NXP
```
Comment (otavio) [inline recipes-bsp/imx-sc-firmware/imx-sc-firmware_1.5.0.bb:2]: Please add (C) and also adjust to 2020.

File: recipes-bsp/imx-uuc/imx-uuc_git.bb [hunk 3]
```diff
@@ -12,7 +12,7 @@ PR = "r1"
 PV = "0.5.1+git${SRCPV}"
 
 SRC_URI = "git://github.com/NXPmicro/imx-uuc.git;protocol=https"
-SRCREV = "ebb7d34853133d92eff9766e370ea21e6d6d8f2b"
+SRCREV = "d6afb27e55d73d7ad08cd2dd51c784d8ec9694dc"
```
Comment (otavio) [inline recipes-bsp/imx-uuc/imx-uuc_git.bb:15]: This commit should be sent as another PR as it can go in right now.

File: recipes-bsp/imx-mkimage/imx-boot_1.0.bb [hunk 4]
```diff
@@ -1,6 +1,7 @@
-# Copyright 2017-2019 NXP
+# Copyright 2017-2020 NXP
```
Comment (otavio) [inline recipes-bsp/imx-mkimage/imx-boot_1.0.bb:1]: Add missed (C)

File: recipes-bsp/imx-mkimage/imx-boot_1.0.bb [hunk 5]
```diff
@@ -1,6 +1,7 @@
...
 
 require imx-mkimage_git.inc
+require recipes-bsp/imx-seco/imx-seco.inc
```
Comment (otavio) [inline recipes-bsp/imx-mkimage/imx-boot_1.0.bb:4]: Why this?

File: recipes-bsp/imx-seco/imx-seco.inc [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+# therefore they are extracted into a separate include file.
+#
+# NOTE: SECO_FIRMWARE_NAME defaults to UNSUPPORTED, which breaks
+# the build if the machine does not have a SECO support.
+SECO_FIRMWARE_NAME              ?= "UNSUPPORTED"
+SECO_FIRMWARE_NAME_mx8qm         = "mx8qmb0-ahab-container.img"
+SECO_FIRMWARE_NAME_mx8qxp        = "mx8qxb0-ahab-container.img"
+SECO_FIRMWARE_NAME_mx8qxpc0      = "mx8qxc0-ahab-container.img"
+SECO_FIRMWARE_NAME_mx8phantomdxl = "mx8qxb0-ahab-container.img"
+SECO_FIRMWARE_NAME_mx8dxl        = "mx8dxla0-ahab-container.img"
```
Comment (otavio) [inline recipes-bsp/imx-seco/imx-seco.inc:11]: Please move this to imx-base as this can be defined on the SoC basis and avoid this to be on the recipe.

File: recipes-bsp/imx-vpu/imx-vpu_5.4.39.3.bb [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+LIC_FILES_CHKSUM = "file://COPYING;md5=228c72f2a91452b8a03c4cab30f30ef9"
+
+PROVIDES = "virtual/imxvpu"
+RPROVIDES_${PN} = "virtual/imxvpu"
+
+PE = "1"
+
+SRC_URI = "${FSL_MIRROR}/${BP}.bin;fsl-eula=true"
+SRC_URI[md5sum] = "6d6302189a6704874375afe62a65def0"
+SRC_URI[sha256sum] = "87cb799a57df654db29403cb74a75ca5185a1517022d3a4a16b8d69056c36127"
```
Comment (otavio) [inline recipes-bsp/imx-vpu/imx-vpu_5.4.39.3.bb:16]: @thochstein can you get the changelog for us?

File: recipes-bsp/imx-vpu-hantro/imx-vpu-hantro_1.18.0.bb [hunk 8]
```diff
@@ -2,22 +2,23 @@
...
 
 PROVIDES = "virtual/imxvpu"
 
 SRC_URI = " \
     ${FSL_MIRROR}/${BPN}-${PV}.bin;fsl-eula=true \
 "
-SRC_URI[md5sum] = "79ca586c67ea9b007fa772df8959c852"
-SRC_URI[sha256sum] = "1f373f5d9a37754f7936f9bf4b134b65afe98459c0383b63c50c1f6ecba49e25"
+SRC_URI[md5sum] = "78034de7ed74363eb793d29894bba5e3"
+SRC_URI[sha256sum] = "bebd82649d00d6dd8236b77b8677b1cc6ac46dc474200502df7797a75dc8f568"
```
Comment (otavio) [inline recipes-bsp/imx-vpu-hantro/imx-vpu-hantro_1.18.0.bb:13]: @thochstein here as well, please.

File: conf/machine/imx8mmevk.conf [hunk 9]
```diff
@@ -39,7 +39,7 @@ SPL_BINARY = "spl/u-boot-spl.bin"
 DDR_FIRMWARE_NAME = "lpddr4_pmu_train_1d_imem.bin lpddr4_pmu_train_1d_dmem.bin lpddr4_pmu_train_2d_imem.bin lpddr4_pmu_train_2d_dmem.bin"
 
 # Set u-boot DTB
-UBOOT_DTB_NAME = "fsl-imx8mm-evk.dtb"
+UBOOT_DTB_NAME = "imx8mm-evk.dtb"
```
Comment (otavio) [inline conf/machine/imx8mmevk.conf:42]: Please squash this on the U-Boot upgrade, as it is required since its update.

File: recipes-kernel/kernel-modules/kernel-module-imx-gpu-viv_6.4.0.p2.4.bb [hunk 10]
```diff
@@ -7,14 +7,14 @@ allowing flexibility to use a newer graphics release with an older kernel."
...
-SRCBRANCH = "lf-5.4.y"
+SRCBRANCH = "imx_5.4.24_2.1.0"
 LOCALVERSION = "-${SRCBRANCH}"
 KERNEL_SRC ?= "git://source.codeaurora.org/external/imx/linux-imx.git;protocol=https"
 SRC_URI = " \
     ${KERNEL_SRC};branch=${SRCBRANCH};subpath=drivers/mxc/gpu-viv;destsuffix=git/src \
     file://Add-makefile.patch \
 "
-SRCREV = "f8118585ee3c7025265b28985fdfe0af96a84466"
+SRCREV = "2f6f3e92a82e51fe39b53ee758273bef26964836"
```
Comment (otavio) [inline recipes-kernel/kernel-modules/kernel-module-imx-gpu-viv_6.4.0.p2.4.bb:17]: @gibsson can you handle our kernel module fork?

File: recipes-bsp/u-boot/u-boot-imx-common.inc (hunk 1 above)
Comment (zandrey) [reply recipes-bsp/u-boot/u-boot-imx-common.inc:8]: Done

File: recipes-bsp/imx-sc-firmware/imx-sc-firmware_1.5.0.bb (hunk 2 above)
Comment (zandrey) [reply recipes-bsp/imx-sc-firmware/imx-sc-firmware_1.5.0.bb]: I went and actually adjusted all *Copyright* notes in recipes I touched so far.

File: recipes-bsp/imx-uuc/imx-uuc_git.bb (hunk 3 above)
Comment (zandrey) [reply recipes-bsp/imx-uuc/imx-uuc_git.bb]: I've dropped it, would send a patch in a separate PR against `master`

File: recipes-bsp/imx-seco/imx-seco.inc (hunk 6 above)
Comment (zandrey) [reply recipes-bsp/imx-seco/imx-seco.inc]: This header in now dropped, all defines are moved to `imx-base.inc`

File: conf/machine/imx8mmevk.conf (hunk 9 above)
Comment (zandrey) [reply conf/machine/imx8mmevk.conf:42]: Done

**Comment by zandrey** [comment]
> 
> 
> @zandrey thank you very much for this. It is of great help :-)

No problem! I hope that some people would benefit from this upgrade as well! 😉 
> 
> I made a few comments and also mentioned a few people on some comments; please go over them and let me know if there is anything you disagree.

(I think) I got them all covered, and updated the branch already.

File: conf/machine/include/imx-base.inc [hunk 11]
```diff
@@ -234,6 +234,18 @@ MACHINE_FIRMWARE_append_use-mainline-bsp = " linux-firmware-imx-sdma-imx6q linux
...
+# therefore they are extracted into a separate include file.
+#
+# NOTE: SECO_FIRMWARE_NAME defaults to UNSUPPORTED, which breaks
+# the build if the machine does not have a SECO support.
+SECO_FIRMWARE_NAME              ?= "UNSUPPORTED"
+SECO_FIRMWARE_NAME_mx8qm         = "mx8qmb0-ahab-container.img"
+SECO_FIRMWARE_NAME_mx8qxp        = "mx8qxb0-ahab-container.img"
+SECO_FIRMWARE_NAME_mx8qxpc0      = "mx8qxc0-ahab-container.img"
+SECO_FIRMWARE_NAME_mx8phantomdxl = "mx8qxb0-ahab-container.img"
+SECO_FIRMWARE_NAME_mx8dxl        = "mx8dxla0-ahab-container.img"
```
Comment (otavio) [inline conf/machine/include/imx-base.inc:247]:
```suggestion
# Below SECO names are used by imx-boot and imx-seco recipes.
SECO_FIRMWARE_NAME              ?= ""
SECO_FIRMWARE_NAME_mx8qm         = "mx8qmb0-ahab-container.img"
SECO_FIRMWARE_NAME_mx8qxp        = "mx8qxb0-ahab-container.img"
SECO_FIRMWARE_NAME_mx8qxpc0      = "mx8qxc0-ahab-container.img"
SECO_FIRMWARE_NAME_mx8phantomdxl = "mx8qxb0-ahab-container.img"
SECO_FIRMWARE_NAME_mx8dxl        = "mx8dxla0-ahab-container.img"
```

Comment (otavio) [reply conf/machine/include/imx-base.inc:247]: On both recipes using those, we should add something like:

```python
python () {
    seco_firmware = d.getVar("SECO_FIRMWARE_NAME", False)
    if not seco_firmware:
        raise bb.parse.SkipRecipe("The 'SECO_FIRMWARE_NAME' must be defined for the SoC in 'imx-base.inc'.")
}
```

I think it does what we need.

**Comment by zandrey** [comment]
> 
> 
> I made some more comments; considering the seco-firmware is being used in multiple places, we might move it to a class so the Python function can be shared as well. What about a bbclass as `use-seco-firmware.bbclass`?

Yes, this makes total sense! I'd make a separate class and re-work recipes accordingly.

Your suggestion would also be considered in the update - thanks for that!

File: recipes-kernel/kernel-modules/kernel-module-imx-gpu-viv_6.4.0.p2.4.bb (hunk 10 above)
Comment (gibsson) [reply recipes-kernel/kernel-modules/kernel-module-imx-gpu-viv_6.4.0.p2.4.bb:17]: @otavio yes will do!

File: recipes-bsp/imx-vpu-hantro/imx-vpu-hantro_1.18.0.bb (hunk 8 above)
Comment (thochstein) [reply recipes-bsp/imx-vpu-hantro/imx-vpu-hantro_1.18.0.bb:13]: It was a bit of work to figure this out, so if you wouldn't mind including the command line and the SHA info, it might help the next time:
```
$ git log --oneline --no-decorate 904a9ab..ee99dc1
ee99dc1 NVJM850-82 - [RV8] RV8 video can't be decoded
b2aa5a0 M865SW-147: [G1/G2]: enhancement for h264 error handing
cfffc5f MMIOT-418: Allow to configure CFG_SECURE_DATA_PATH on Linux
0a741cc NVJM850-81 - [G2]:hevc error handling
2f3f4fa M865SW-132: [G1G2]: HEVC decoder illegal memory access in security mode
8182f16 MA-14108-1 Remove ION_FLAG_CACHED flag for hantro vpu buffer
65d5f3a MMFMWK-8595 Enable VC8000E encoder for 865
b434f22 M865SW-87:[G1/G2]: Integrate some code for VSI platform
4dc2d6f M865SW-82 - [G1/G2]: H264 conformance test failure caused by change in 1.0.6 release
dd606d2 [G1/G2]: Integrate VSI 1.0.7_20191212 release
d62fd74 [G1/G2]: Coverity issues fix
6460a5f [G1/G2]: Refine Test Makefile
7805f7f [G1/G2]: Coverity issues fix
eca7409 [G1/G2]: Coverity issues fix
a210f04 [G1/G2]: Coverity issues fix
dd0e5f4 [G1/G2]: Coverity issues fix
c12e5a8 [G1/G2]: Coverity issues fix
7631926 [G1/G2]: Coverity issues fix
```

File: recipes-bsp/imx-vpu/imx-vpu_5.4.39.3.bb (hunk 7 above)
Comment (thochstein) [reply recipes-bsp/imx-vpu/imx-vpu_5.4.39.3.bb:16]:
```
$ git log --oneline --no-decorate 3e47ffe..d8f988d
d8f988d MA-14108 Remove ION_FLAG_CACHED flag for cnm vpu buffer
```

**Comment by zandrey** [comment]
@otavio I've worked out all the latest comments from your side, `u-boot-imx` component upgrade in now combined into a single commit.

As for the `SECO`, I've extracted the naming logic into a separate class. The only thing I had to address there in addition is to sort out the `MACHINEOVERRIDES` since without having an additional check there for `mx8x` family of SoCs - all others (e.g. `mx8m`) would be failing to build as they include the `imx-boot` recipe but do not define `SECO_FIRMWARE_NAME`.

Please have a look at it and tell me if that is OK to go like this.

@thochstein Thanks for the info regarding commits that went into the upgraded versions, I've included those into commit message (together with command line as you've asked) and put you there as `Cc:` in case if a reference later would be needed.

-- andrey

File: recipes-kernel/kernel-modules/kernel-module-imx-gpu-viv_6.4.0.p2.4.bb (hunk 10 above)
Comment (gibsson) [reply recipes-kernel/kernel-modules/kernel-module-imx-gpu-viv_6.4.0.p2.4.bb:17]: @otavio: I split it in 2 patches: https://github.com/Freescale/meta-freescale/pull/462
The first one, update to 6.4.0.p2.0 should be backported to dunfell whereas the second one should only be on master.
Hope it makes sense.

File: classes/use-imx-seco-firmware.bbclass [hunk 12]
```diff
@@ -0,0 +1,32 @@
+#
+# Class to provide a possibility to re-use names of SECO Firmware
```
Comment (thochstein) [inline classes/use-imx-seco-firmware.bbclass:2]: Suggest to expand the acronym at the beginning to SEcurity COntroller.

File: classes/use-imx-seco-firmware.bbclass [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+SECO_FIRMWARE_NAME              ?= ""
+
+SECO_FIRMWARE_NAME_mx8qm         = "mx8qmb0-ahab-container.img"
+SECO_FIRMWARE_NAME_mx8qxp        = "mx8qxb0-ahab-container.img"
+SECO_FIRMWARE_NAME_mx8qxpc0      = "mx8qxc0-ahab-container.img"
+SECO_FIRMWARE_NAME_mx8phantomdxl = "mx8qxb0-ahab-container.img"
+SECO_FIRMWARE_NAME_mx8dxl        = "mx8dxla0-ahab-container.img"
+
+python () {
+    if "mx8x" in d.getVar('MACHINEOVERRIDES').split(":"):
```
Comment (thochstein) [inline classes/use-imx-seco-firmware.bbclass:28]: The firmware is valid for both 8 and 8X families.

File: classes/use-imx-seco-firmware.bbclass (hunk 12 above)
Comment (zandrey) [reply classes/use-imx-seco-firmware.bbclass:2]: Good point, would expand it.

File: classes/use-imx-seco-firmware.bbclass (hunk 13 above)
Comment (zandrey) [reply classes/use-imx-seco-firmware.bbclass:28]: Does that mean that only 8M family have no SECO FW required?

In other words, would it be safe to put something like:
```python

python () {
    if "mx8m" not in d.getVar('MACHINEOVERRIDES').split(":"):
        seco_firmware = d.getVar('SECO_FIRMWARE_NAME')
        if not seco_firmware:
            raise bb.parse.SkipRecipe("This SoC requires 'SECO_FIRMWARE_NAME', define it in 'use-imx-seco-firmware' bbclass")
}

```

Comment (thochstein) [reply classes/use-imx-seco-firmware.bbclass:28]: Yes, that will work.

Comment (zandrey) [reply classes/use-imx-seco-firmware.bbclass:28]: Thanks for the update! I've tried it out and it seems to be working for derivatives of 8 and 8X families.

I would make an adjustment and push an updated version.

File: classes/use-imx-seco-firmware.bbclass [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+        seco_firmware = d.getVar('SECO_FIRMWARE_NAME')
+        if not seco_firmware:
+            raise bb.parse.SkipRecipe("This SoC requires 'SECO_FIRMWARE_NAME', define it in 'use-imx-seco-firmware' bbclass")
+}
```
Comment (otavio) [inline classes/use-imx-seco-firmware.bbclass:32]: I believe we don't need to check the `mx8x` on `MACHINEOVERRIDES` as it'll fail if we don't have the firmware on the list. So it is redundant.

File: recipes-bsp/imx-seco/imx-seco_3.6.3.bb [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+do_install[noexec] = "1"
+
+addtask deploy after do_install
+do_deploy () {
+    # Deploy i.MX8 SECO firmware files
+    install -m 0644 ${S}/firmware/seco/${SECO_FIRMWARE_NAME} ${DEPLOYDIR}
+}
+
+COMPATIBLE_MACHINE = "(mx8)"
+COMPATIBLE_MACHINE_mx8m = "(^$)"
```
Comment (otavio) [inline recipes-bsp/imx-seco/imx-seco_3.6.3.bb:27]: The compatible machine ought to be moved to the class. Please also use the expanded list and set the `PACKAGE_ARCH` to `MACHINE_SOC_ARCH` so it is reused for all recipes.

File: classes/use-imx-seco-firmware.bbclass (hunk 14 above)
Comment (zandrey) [reply classes/use-imx-seco-firmware.bbclass:32]: I already made a suggestion above based on comment from @thochstein that filter would instead be looking at all overrides but `mx8m` where SECO FW is not required. Without that in place `imx-boot` fails to build for `mx8m` derivatives since the `SECO_FIRMWARE_NAME` is set to empty string.

Would it be OK to adjust the parsing filter like I suggested above?

File: classes/use-imx-seco-firmware.bbclass (hunk 13 above)
Comment (otavio) [reply classes/use-imx-seco-firmware.bbclass:28]: Possibly like this:

```python
python () {
    seco_firmware = d.getVar('SECO_FIRMWARE_NAME')
    if not seco_firmware and "mx8m" not in d.getVar('MACHINEOVERRIDES').split(":"):
        raise bb.parse.SkipRecipe("This SoC requires 'SECO_FIRMWARE_NAME', define it in 'use-imx-seco-firmware' bbclass")
}
```

**Comment by otavio** [comment]
I had a good talk with @zandrey and we agreed on some rework for the PR. Some ready parts will be moved to master soon and reduce the size of the PR :-)

**Comment by zandrey** [comment]
I've updated the PR again to incorporate (as I think) all the review comment above.

The only thing that is left unattended at the moment is the `imx-boot` recipe, which requires a major re-work to be done: there is a need to have a better mechanism to collect all binary files to pass it over for assembly into the final boot binary. As this appears to be quite an effort - I suggest that this is to be addressed separately.

@otavio @gibsson @thochstein Please have a look at the updated version and tell me if it's good to go.

-- andrey
